### PR TITLE
Make field limit more predictable

### PR DESCRIPTION
**Pull request by felixbarny**
Today, we're counting all mappers, including mappers for subfields that aren't explicitly added to the mapping towards the field limit.

This means that some field types, such as `search_as_you_type` or `percolator` count as more than one field even though that's not apparent to users as they're just defining them as a single field in the mapping.

This change makes it so that each field mapper only counts as one. We're still counting multi-fields.

This makes it easier to understand for users why the field limit is hit.

~In addition to that, it also simplifies https://github.com/elastic/elasticsearch/pull/96235 as it makes the implementation of `Mapper.Builder#getTotalFieldsCount` much easier and easier to align with `Mapper#getTotalFieldsCount`. This reduces the risk of over- or under-estimating the field count of a `Mapper.Builder` in `DocumentParserContext#addDynamicMapper`,
which in turn reduces the risk of data loss due to the issue described here: https://github.com/elastic/elasticsearch/pull/96235#discussion_r1402495749.~

*Edit: due to https://github.com/elastic/elasticsearch/pull/103865, we don't need an implementation of `getTotalFieldsCount` or `mapperSize` in `Mapper.Builder`. Still, this PR more closely aligns `Mapper#getTotalFieldsCount` with `MappingLookup#getTotalFieldsCount`, which  `DocumentParserContext#addDynamicMapper` uses to determine whether the field limit is hit*

A potential risk of this is that we're now effectively allowing more fields in the mapping. It may be surprising to users that more fields can be added to a mapping. Although, I'd not expect negative consequences from that. Generally, I'd  expect users to be happy about any change that reduces the risk of data loss.

We could also think about whether to apply the new counting logic only to new indices (depending on the `IndexVersion`). However, that would add more complexity and I'm not convinced about the value. We'd then need to maintain two different ways of counting fields and also require passing in the `IndexVersion` to `MappingLookup` which previously didn't require the `IndexVersion`.

This PR is meant as a conversation starter. It would also simplify https://github.com/elastic/elasticsearch/pull/96235 but I don't think this blocks that PR in any way.

I'm curious about the opinion of @javanna and @jpountz on this.

### CODE REVIEW

**Comment by jpountz** [comment]
+1 to this change. The thing that is not entirely clear to me is what should we do about meta fields (`_id` , `_source`, etc.). It might be better to not count them either?

There is a strong benefit for us, which is that recording additional fields under the hood is no longer a change that affects users, if we don't count these fields.

**Comment by felixbarny** [comment]
Meta fields (`Mapping#metadataMappers`) haven't been counted previously and they also aren't counted with this PR (only `Mapping#root`).

**Comment by felixbarny** [comment]
> There is a strong benefit for us, which is that recording additional fields under the hood is no longer a change that affects users, if we don't count these fields.

I think this is a good point. Currently, if we'd decide to add additional mappers for a field, that might be seen as a breaking change as users can then add fewer of these fields to the mapping. So this change would avoid these kinds of breaking changes in the future.

**Comment by salvatore-campagna** [comment]
LGTM

**Review by salvatore-campagna — APPROVED** [review]

File: server/src/main/java/org/elasticsearch/index/mapper/MappingLookup.java [hunk 1]
```diff
@@ -223,6 +225,14 @@ FieldTypeLookup fieldTypesLookup() {
...
      */
     public long getTotalFieldsCount() {
+        return totalFieldsCount;
+    }
+
+    /**
+     * Returns the total number of mappers defined in the mappings, including field mappers and their sub-fields
+     * (which are not explicitly defined in the mappings), multi-fields, object mappers, runtime fields and metadata field mappers.
+     */
+    public long getTotalMapperCount() {
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/index/mapper/MappingLookup.java:235]: I think I can no longer tell the difference between getTotalFieldsCount and getTotalMapperCount . It's very subtle, or is there any difference at all?

**Review by javanna — COMMENTED** [review]
I agree that we should not complicate things. If all this change does is possibly allow for more fields in the mappings with the same total fields limit, it should not be a concern. It is more permissive and won't break existing users. It does mean that docs/mappings that were rejected before the upgrade are possibly going to be accepted after the upgrade, for the same indices, which is slightly weird. I am not convinced though that we should tie this change to the index created version.

I must admit that I still find it difficult to summarize what counting this change affects, in that it moves things around by delegating the counting to each mapper, and I am not sure what part of that is just mechanical changes and what part affects the actual counting semantics. Is the only difference those special cases like percolator and search as you type, which have more fields than one, and will now count as 1, while before we were not able to make that distinction within MappingLookup?

Should we update the docs on this or perhaps it is too much in the weeds that users don't need to know? I guess we don't even document the current behaviour that clearly?

**Comment by felixbarny** [comment]
> Is the only difference those special cases like percolator and search as you type, which have more fields than one, and will now count as 1, while before we were not able to make that distinction within MappingLookup?

Yes, this pretty much summarizes the change. Another way to put it would be to say that now we're counting only the fields that are visible in the mapping. Before the change, we counted mappers, not fields and a field may have multiple mappers.

This is also why `MappingLookup` now has two different methods: `getTotalMapperCount` returns the number of mapper instances in the mapping, which is used to estimate the storage overhead of the mapping in `NodeMappingStats`. The `getTotalFieldsCount` method is used when it comes to validating the field limit or checking whether we have budget left to add a dynamic field.

> Should we update the docs on this or perhaps it is too much in the weeds that users don't need to know? I guess we don't even document the current behaviour that clearly?

I don't feel like docs changes are needed. It should be clearer and more easy to reason about why the field limit has been hit than before. Now, users can just count the fields in the mapping and when the limit is hit, we'd expect that the number of fields visible in the mapping is equal to the field limit.

File: server/src/main/java/org/elasticsearch/index/mapper/MappingLookup.java (hunk 1 above)
Comment (felixbarny) [reply server/src/main/java/org/elasticsearch/index/mapper/MappingLookup.java:235]: See https://github.com/elastic/elasticsearch/pull/102885#issuecomment-1929207955

**Review by javanna — APPROVED** [review]
LGTM

**Pull request merged**